python/cudf/cudf/core/udf/groupby_utils.py

```diff
@@ -0,0 +1,181 @@
+    See numba.np.numpy_support.from_struct_dtype for details.
+    """
+
+    # Create the numpy structured type corresponding to the numpy dtype.
+    fields = []
+    offset = 0
+
+    sizes = [val[0].itemsize for val in dtype.fields.values()]
+    for i, (name, info) in enumerate(dtype.fields.items()):
+        elemdtype = info[0]
+        title = info[2] if len(info) == 3 else None
+        ty = numpy_support.from_dtype(elemdtype)
+        indexty = numpy_support.from_dtype(index_dtype)
+        infos = {
+            "type": GroupType(ty, indexty),
+            "offset": offset,
+            "title": title,
+        }
+        fields.append((name, infos))
+
+        offset += 8 + 8 + 8  # group struct size (2 pointers and 1 integer)
+
+        # Align the next member of the struct to be a multiple of the
+        # memory access size, per PTX ISA 7.4/5.4.5
+        if i < len(sizes) - 1:
+            # next_itemsize = sizes[i + 1]
+            next_itemsize = 8
+            offset = int(math.ceil(offset / next_itemsize) * next_itemsize)
+
+    # Numba requires that structures are aligned for the CUDA target
+    _is_aligned_struct = True
+    return Record(fields, offset, _is_aligned_struct)
+
+
+def _groupby_apply_kernel_string_from_template(frame, args):
+    """
+    Function to write numba kernels for `Groupby.apply` as a string.
+    Workaround until numba supports functions that use `*args`
+    """
+    # Create argument list for kernel
+    frame = _supported_cols_from_frame(
+        frame, supported_types=SUPPORTED_GROUPBY_NUMPY_TYPES
+    )
+    input_columns = ", ".join([f"input_col_{i}" for i in range(len(frame))])
+    extra_args = ", ".join([f"extra_arg_{i}" for i in range(len(args))])
+
+    # Generate the initializers for each device function argument
+    initializers = []
+    for i, colname in enumerate(frame.keys()):
+        initializers.append(
+            group_initializer_template.format(idx=i, name=colname)
+        )
+
+    return groupby_apply_kernel_template.format(
+        input_columns=input_columns,
+        extra_args=extra_args,
+        group_initializers="\n".join(initializers),
+    )
+
+
+def _get_groupby_apply_kernel(frame, func, args):
+    np_field_types = np.dtype(
+        list(
+            _supported_dtypes_from_frame(
+                frame, supported_types=SUPPORTED_GROUPBY_NUMPY_TYPES
+            ).items()
+        )
+    )
+    dataframe_group_type = _get_frame_groupby_type(
+        np_field_types, frame.index.dtype
+    )
+    return_type = _get_udf_return_type(dataframe_group_type, func, args)
+
+    # Dict of 'local' variables into which `_kernel` is defined
+    global_exec_context = {
+        "cuda": cuda,
+        "Group": Group,
+        "dataframe_group_type": dataframe_group_type,
+        "types": types,
+    }
+    kernel_string = _groupby_apply_kernel_string_from_template(frame, args)
+
+    kernel = _get_kernel(kernel_string, global_exec_context, None, func)
```

Comment:
So this compiles the "string" python code by `exec`ing it and then calling numba.cuda.jit?

I would love a way to avoid string-based metaprogramming, but I guess that is tricky.

Reply:
Yup. This is also my least favorite part of `apply`.